pypesto/objective/history.py

```diff
@@ -709,8 +903,38 @@ def _update_vals(self,
                 self.res_min = result.get(RES)
                 self.sres_min = result.get(SRES)
 
-
-def ndarray2string_full(x: np.ndarray):
+    def _compute_vals_from_trace(self):
+        # some optimizers may evaluate hess+grad first to compute trust region
+        # etc
+        for it in range(3):
+            fval0_candidate = self.history.get_fval(it)
+            if not np.isnan(fval0_candidate) \
+                    and np.allclose(self.history.get_x(it), self.x0):
+                self.fval0 = fval0_candidate
+
+        iter_min = np.nanargmin(self.history.get_fval_trace())
+        if not isinstance(iter_min, np.int64):
+            iter_min = iter_min[0]
+
+        self.fval_min = self.history.get_fval(iter_min)
+        self.x_min = self.history.get_x(iter_min)
+        if self.history.options.trace_record_grad:
+            self.grad_min = self.history.get_grad(iter_min)
+            if np.isnan(self.grad_min).all() and iter_min < len(self.history):
+                # gradient typically evaluated on the next call
```

Comment:
this can be wrong, right? in that case maybe issue a warning that the gradient may not be correct?

Reply:
fixed by ensuring that x is same on next iteration, added doc